src/managedwriter/stream_connection.ts

```diff
@@ -87,6 +86,18 @@ export class StreamConnection extends EventEmitter {
     this._connection.on('error', this.handleError);
     this._connection.on('close', () => {
       this.trace('connection closed');
+      if (this.hasPendingWrites()) {
+        this.reconnect();
```

Comment:
does StreamConnection have a notion of terminal error where we wouldn't want reconnect to happen?  I see there's some notion of permanent error further down, but I may not be grokking the flow here.

Reply:
Yeah, I think it makes sense to only reconnect when there is pending requests && it is retrable error.

Reply:
the problem is that the 'close' callback doesn't give the error that caused it. I can create an instance variable on keep track of the last seen error, but not sure how safe this is.

Reply:
ouch...
I think it is still safe here, if we reconnect on permenant error and resend all the request, the connection will be shutdown, but all the pending requests will eventually fail with a limited number of retries. So the pending queue will eventually drain. Also, I think it is worth a logging here, just say we are reconnecting with X number of pending writes.